### ISSUE TITLE
Potential optimizer problem in bytes32 shift

### ISSUE THREAD

**Issue by nweller** [ethereum/solidity#5116]
Using truffle and ganache-cli (ethereumjs-testrpc-sc 6.1.6) with various binary solc as well as solcjs versions including the recent 0.4.25 release, enabling optimization in truffle with:

    solc: {
      optimizer: {
        enabled: true,
        runs:  0 
      },

causes the test case below to produce this ganache-cli error, which terminates the test:

    eth_sendTransaction
    Error: redPow(normalNum)
        at r (/home/nweller/blockchain-compiler-testing/test-env/test-env-truffle/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:2:20678)
        at c.redPow (/home/nweller/blockchain-compiler-testing/test-env/test-env-truffle/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:2:57378)
        at EXP (/home/nweller/blockchain-compiler-testing/test-env/test-env-truffle/node_modules/ethereumjs-testrpc-sc/build/cli.node.js:52:1883988)

The test case - which requires a transaction with f0() to fail - is this:

    contract c0 {
        uint184 v1 = uint184(1); 

        // If instead of uint184 we use a smaller type like uint8 or uint16, the problem disappears:
        // uint16 v1 = uint16(1); 

        bytes32 v2 = bytes32(0x02); 

        function f0() public payable {
            // FAILURE: The shift operation causes a ganache-cli error
            emit BYTES32(v2 >> v1);
        }

        event BYTES32(bytes32 arg);
    }

Unfortunately, I cannot reproduce this in Remix.

Does it look like a compiler optimization problem, or is something off with ganache-cli? I have lots of test cases that fail in this manner when I enable optimization. If you have any advice on how to avoid known optimizer problems for test case generation, please let me know.

(This report is part of an ICE Center@ETH Zurich project on automated compiler validation funded by the Ethereum Foundation.)

**Comment by axic** [comment]
Both ganache and Remix use ethereumjs-vm and this seems to be a problem with ganache using an old version. `redPow` is used internally to do exponentiation, and on pre-constantinople EVM, shifts are done using a combination of multiplication/division and exponentiation.

While both of these tools are good to get some initial idea if something might be broken, I would not at all rely on their accuracy of implementing the EVM specification.

The best is to use a consensus passing client with Remix or Truffle or just manually examine the generated EVM bytecode by Solidity.

Alternatively one could add a test case to `tests/libsolidity/SolidityEndToEnd.cpp` which will be run through aleth, another consensus passing client.

**Comment by nweller** [comment]
Thank you very much for the information, Alex.  I can confirm that ethereumjs-testrpc-sc appears to use an outdated ethereumjs-vm fork (already found another case - "Error: Already a number in reduction context" - , which cannot be reproduced in Remix either). I'll take it to the truffle / ganache-cli developers. 

I've tried your suggestion of getting started with aleth, but this hasn't worked for me so far. I asked in various gitter chats, but haven't received much of a response apart from: it's an unusual combination which may be unlikely to work.

I got aleth and truffle to exchange a few RPC messages, but then truffle fails with: "Error: Don't set config.from directly. Instead, set config.networks and then config.networks[<network name>].from"

Guess I may have to debug and fix it myself or give up on optimization tests for now.

Any word from you on truffle's decision to disable optimization by default due to previous issues: https://github.com/trufflesuite/truffle-compile/pull/5 ?

Is this negative stance on optimization blown out of proportion / outdated?
